src/advanced_inv.cpp

```diff
@@ -760,49 +797,37 @@ bool advanced_inventory_pane::is_filtered( const item *it ) const
     return filtercache[ str ];
 }
 
-template <typename Container>
-static itemslice i_stacked( Container items )
+// roll our own, to handle moving stacks better
+typedef std::vector<std::list<item*>> itemstack;
+
+template <typename T>
+static itemstack i_stacked(T items)
 {
     //create a new container for our stacked items
-    itemslice islice;
-
-    //iterate through all items in the vector
-    for( auto &items_it : items ) {
-        if( items_it.count_by_charges() ) {
-            // Those exists as a single item all the item anyway
-            islice.push_back( std::make_pair( &items_it, 1 ) );
-            continue;
-        }
-        bool list_exists = false;
-
-        //iterate through stacked item lists
-        for( auto &elem : islice ) {
-            //check if the ID exists
-            item *first_item = elem.first;
-            if( first_item->type->id == items_it.type->id ) {
-                //we've found the list of items with the same type ID
-
-                if( first_item->stacks_with( items_it ) ) {
-                    //add it to the existing list
-                    elem.second++;
-                    list_exists = true;
-                    break;
-                }
+    itemstack stacks;
+    // used to recall index we stored `itype_id' item at in itemstack
+    std::unordered_map<itype_id, int> cache;
```

Comment:
So I changed it to test for each element in the stack, as opposed to just the front element, so that should hopefully fix the stacking of items with contents.